Extending modules remodel (#905)
ezsystems/developer-documentation: docs/guide/extending/extending_udw.md

> # Extending Universal Discovery Widget (UDW)

Reviewer: ```suggestion
# Extending Universal Discovery Widget
```

> Universal Discovery module allows you to browse the content structure and search for content

Reviewer: ```suggestion
Universal Discovery Widget (UDW) allows you to browse the content structure and search for content
```

using an interactive interface: browse, search, create, and bookmarks view .

## How to use UDW?

With vanilla JS:

``` js
const container = document.querySelector('#react-udw');

ReactDOM.render(React.createElement(eZ.modules.UniversalDiscovery, {
    restInfo: {
        token: {String},
        siteaccess: {String}
    }
    onConfirm: {Function},
    onCancel: {Function},
}), container);
```

With JSX:

``` jsx
const props = {
    onConfirm: {Function},
    onCancel: {Function}
};

<UniversalDiscoveryModule {...props} />
```

## Adding new tabs to the UDW

> The Universal Discovery module is highly customizable. It allows you to add new tabs to the module.

Reviewer: ```suggestion
The Universal Discovery Widget is highly customizable. It allows you to add new tabs to the module.
```

    - **isVisible** _{Boolean}_ - visible flag
    - **onItemSelect** _{Function}_ - a callback to be invoked when content is selected
    - **maxHeight** _{Number}_ - the maximum height of the panel container
    - **id** _{String}_ - panel identifier
    - **startingLocationId** _{Number}_ - Location ID
    - **findLocationsByParentLocationId** _{Function}_ - finds Locations related to the parent Location
    - **findContentBySearchQuery** _{Function}_ - finds content matching a given text query
    - **contentTypesMap** _{Object}_ - Content Type map with Content Type IDs as keys
    - **multiple** _{Boolean}_ - can select multiple Content items flag
    - **labels** _{Object}_ - a hash containing text messages to be placed across many places in a component
- **attrs** {Object} - any optional list of props that should applied to the panel component.
It can override the panel props listed above.

## Property list

The `<UniversalDiscoveryModule />` module can handle additional properties.
There are 2 types of properties: **required** and **optional**.

### Required properties

> Without all the following properties the Universal Discovery module will not work.

Reviewer: ```suggestion
Without all the following properties the Universal Discovery Widget will not work.
```

**onConfirm** _{Function}_ - a callback to be invoked when a user clicks on the confirm button
in a Universal Discovery popup. The function takes one param: `content` which is an array of Content item structs.

**onCancel** _{Function}_ - a callback to be invoked when a user clicks on the cancel button
in a Universal Discovery popup. It takes no extra params.

**restInfo** _{Function}_ - a config hash containing: token (_{String}_) and SiteAccess (_{String}_).

### Optional props

> Optionally, Universal Discovery module can take a following list of props:

Reviewer: ```suggestion
Optionally, Universal Discovery Widget can take a following list of props:
```

- **selectedItemsLimit** _{Number}_ - the limit of items that can be selected. Should be combined with the `multiple` attribute set to `true`. Default value is `0`, which means no limit,
- **allowContainersOnly** _{Boolean}_ - when true, only containers can be selected. Default value: `false`,
- **cotfPreselectedLanguage** _{String}_ - language that is preselected in Content on the Fly,
- **cotfAllowedLanguages** _{Array}_ - languages that are available in Content on the Fly,
- **cotfPreselectedContentType** _{String}_ - Content Type that is preselected in Content on the Fly,
- **cotfAllowedContentTypes** _{Array}_ - Content Types that are available in Content on the Fly,
- **allowedContentTypes** _{Array}_ - Content Types that are available in other UDW tabs,
- **cotfPreselectedLocation** _{Number}_ - Location that is preselected in Content on the Fly,
- **cotfAllowedLocations** _{Array}_ - Locations that are available in Content on the Fly

The following props are deprecated:

- **languages** and **contentTypes** are lists of languages and Content Types in the system, read from the application config.

## Configuration

> You can configure Universal Discovery module in the [`universal_discovery_widget.yaml`](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/src/bundle/Resources/config/universal_discovery_widget.yaml) file.

Reviewer: ```suggestion
You can configure Universal Discovery Widget in the [`universal_discovery_widget.yaml`](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/src/bundle/Resources/config/universal_discovery_widget.yaml) file.
```

There you can set e.g. the starting location ID, visible tabs, allowed Content Types, search limits, etc.

```yaml
system:
    <siteaccess|siteaccess_group>:
        universal_discovery_widget_module:
            configuration:
                default:
                    starting_location_id: 1
                    visible_tabs: [browse, search, bookmarks]
                    allowed_content_types: []
                    search:
                        results_per_page: 10
                        limit: 50
```

UDW configuration is SiteAccess-aware.
For each defined SiteAccess, you need to be able to use the same configuration tree in order to define SiteAccess-specific config.
These settings need to be mapped to SiteAccess-aware internal parameters that you can retrieve via the ConfigResolver.
For more information on ConfigResolver, see [eZ Platform dynamic configuration basics.](../config_dynamic#configresolver)

### Adding new configuration

UDW configuration can change dynamically depending on occurring events.
It can be used e.g. for defining which content should be exposed to a user after logging in.

> By default only one element from configuration file is applied to Universal Discovery module.

Reviewer: ```suggestion
By default only one element from configuration file is applied to Universal Discovery Widget.
```
